eks/terraform/modules/cluster/irsa.tf

```diff
@@ -0,0 +1,216 @@
+  version = "5.34.0"
+
+  role_name          = "${var.cluster_name}-cluster-autoscaler"
+  policy_name_prefix = "${var.cluster_name}-"
+
+  attach_cluster_autoscaler_policy = true
+  cluster_autoscaler_cluster_ids   = [aws_eks_cluster.cluster.id]
+
+  oidc_providers = {
+    main = {
+      provider_arn               = aws_iam_openid_connect_provider.cluster[0].arn
+      namespace_service_accounts = ["kube-system:${local.cluster_autoscaler_service_account}"]
+    }
+  }
+}
+
+moved {
+  from = module.cluster_autoscaler_irsa_role
+  to   = module.cluster_autoscaler_irsa_role[0]
+}
+
+
+module "loadbalancer_controller_irsa_role" {
+  count = var.use_irsa_v1 ? 1 : 0
+
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
+  version = "5.34.0"
+
+  role_name          = "${var.cluster_name}-loadbalancer-controller"
+  policy_name_prefix = "${var.cluster_name}-"
+
+  attach_load_balancer_controller_policy = true
+
+  oidc_providers = {
+    main = {
+      provider_arn               = aws_iam_openid_connect_provider.cluster[0].arn
+      namespace_service_accounts = ["kube-system:${local.loadbalancer_controller_service_account}"]
+    }
+  }
+}
+
+moved {
+  from = module.loadbalancer_controller_irsa_role
+  to   = module.loadbalancer_controller_irsa_role[0]
+}
+
+module "ebs_csi_irsa_role" {
+  count = var.use_irsa_v1 ? 1 : 0
+
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
+  version = "5.34.0"
+
+  role_name          = "${var.cluster_name}-ebs-csi"
+  policy_name_prefix = "${var.cluster_name}-"
+
+  attach_ebs_csi_policy = true
+
+  oidc_providers = {
+    main = {
+      provider_arn               = aws_iam_openid_connect_provider.cluster[0].arn
+      namespace_service_accounts = ["kube-system:ebs-csi-controller-sa"]
+    }
+  }
+}
+
+moved {
+  from = module.ebs_csi_irsa_role
+  to   = module.ebs_csi_irsa_role[0]
+}
+
+module "vpc_cni_irsa_role" {
+  count = var.use_irsa_v1 ? 1 : 0
+
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
+  version = "5.34.0"
+
+  role_name          = "${var.cluster_name}-vpc-cni"
+  policy_name_prefix = "${var.cluster_name}-"
+
+  attach_vpc_cni_policy = true
+  vpc_cni_enable_ipv4   = true
+
+  oidc_providers = {
+    main = {
+      provider_arn               = aws_iam_openid_connect_provider.cluster[0].arn
+      namespace_service_accounts = ["kube-system:aws-node"]
+    }
+  }
+}
+
+moved {
+  from = module.vpc_cni_irsa_role
+  to   = module.vpc_cni_irsa_role[0]
+}
+
+################################################################################
+# IRSA v1
+################################################################################
+module "cluster_autoscaler_pod_identity" {
+  count = var.use_irsa_v2 ? 1 : 0
+
+  source  = "terraform-aws-modules/eks-pod-identity/aws"
+  version = "1.2.1"
+
+  name = "${var.cluster_name}-ca"
```

Comment:
We need to also set `use_name_prefix` to `false` because by default it uses prefix to create all resources and the prefix is having length limit of 37. Hence, the deployment will fail for long datacenter name. I found the issue when I was testing the pod identity migration for existing cluster.

Reply:
hmm this is the original code that uses irsa, I have to look into how to make it backward compatible.